### PR TITLE
Feature/api/patch contact

### PR DESCRIPTION
**Pull request by sapiens-sapide**
working implementation of Patch verb on route /contacts/{contact_id}, with Caliopen's Patch specifications. To be tested.

### CODE REVIEW

File: src/backend/main/py.main/caliopen_main/user/core/contact.py [hunk 1]
```diff
@@ -297,4 +370,4 @@ def add_public_key(self, key):
         return self._add_relation('public_keys', key)
 
     def delete_public_key(self, key_id):
-        return self._delete_relation('public_keys', key_id)
+        return self._delete_relation('public_keys', key_id)
```
Comment (gdchamal) [inline src/backend/main/py.main/caliopen_main/user/core/contact.py]: missing newline at end of file

File: src/backend/defs/rest-api/paths/contacts.json [hunk 2]
```diff
@@ -208,11 +208,50 @@
...
+            "$ref": "../objects/Error.json"
+          }
+        },
+        "404": {
+          "description": "contact not found",
+          "schema": {
+            "$ref": "../objects/Error.json"
+          }
+        },
+        "422": {
```
Comment (sapiens-sapide) [inline src/backend/defs/rest-api/paths/contacts.json]: done with commit c31f977

File: src/backend/main/py.main/caliopen_main/user/core/contact.py [hunk 3]
```diff
@@ -245,6 +255,69 @@ def lookup(cls, user, value):
...
+
+        # populate patch to a contactParams instance to validate fields against model
+        try:
+            contact_model = params_contact.Contact(patch)
+            contact_model.validate(partial=True)
+        except Exception:
+            return main_errors.PatchUnprocessable(message="Patch does not validate against model")
+
+        # check if patch 'current_state' is consistent with current db state
+        from ..returns import ReturnContact
```
Comment (sapiens-sapide) [inline src/backend/main/py.main/caliopen_main/user/core/contact.py]: Don't know how to compare apples and pears with current object model ! 
New object model proposal with PR #97

File: src/backend/main/py.main/caliopen_main/user/core/contact.py [hunk 4]
```diff
@@ -245,6 +255,69 @@ def lookup(cls, user, value):
...
+                        return main_errors.PatchUnprocessable(message="Patch current_state not consistent with db")
+                elif patch_current_state[k] != contact_db_state[k]:
+                    return main_errors.PatchUnprocessable(message="Patch current_state not consistent with db")
+            else:
+                return main_errors.PatchUnprocessable(message="Invalid key within current_state patch object")
+
+        for key in patch:
+            attr = getattr(contact_model, key)
+            if key in cls._nested:
+                setattr(contact_store, key, cls.create_nested(attr, cls._nested[key]))
```
Comment (sapiens-sapide) [inline src/backend/main/py.main/caliopen_main/user/core/contact.py]: not easy to implement and not sure that cqlengine driver implements correctly the updating of nested objects.

**Comment by gdchamal** [comment]
discarding this MR as #97 address better object representation using zope.interface

**Pull request closed without merging**